### DISCUSSION TITLE
[v7] Issues with readOnly/disabled during form submit

### DISCUSSION THREAD

**Discussion by mogelbrod** [react-hook-form/react-hook-form#4413] [General]
## `readOnly` not a replacement for `disabled`
There has been a number of different questions and issues regarding how to disable input(s) while retaining the value, and the general answer is to use `readOnly` instead of `disabled` for the input element(s):
1. [Best way to get value from disabled <input> #3157](https://github.com/react-hook-form/react-hook-form/discussions/3157)
2. [Disable Fields But Still Return The Data? #2686](https://github.com/react-hook-form/react-hook-form/discussions/2686)
3. [Can disabled input appear in handleSubmit #3936](https://github.com/react-hook-form/react-hook-form/discussions/3936)
4. [Data for disabled fields is not being validated even with a value #2868](https://github.com/react-hook-form/react-hook-form/discussions/2868)

`readOnly` is however only supported by a subset of available HTML inputs - not including checkboxes, radio buttons and select elements, amongst others ([full list and more info at MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/readonly)). This makes it an undesirable choice for a general purpose solution as these input types are pretty common.

I've created [a codesandbox](https://codesandbox.io/s/react-hook-form-readonly-vs-disabled-ei6cs?file=/src/index.js) to showcase this issue. Click on the form submit button, or toggle "simulate submitting" and then try to interact with the various inputs to see how `readOnly` affects them. You can then compare the behaviour with `disabled` by selecting "disabled" in "Input state during submit".

## Disabling the form (inputs) while retaining values
Disabling the form during submission is also a common form behaviour (brought up in [this thread](https://github.com/react-hook-form/react-hook-form/discussions/3024) but never solved). Doing so causes RHF to consider the value to be `undefined` however, to match browser behaviour. This results in RHF watchers and `getValues()` calls returning `undefined` for these values while the form is submitting, which is undesirable.

The workaround I found for this is to **wrap the elements with a `<fieldset>` and disable it during submit rather than the inputs themselves**. This can be tested in the linked CSB by selecting "disabled fieldset" in "Input state during submit". It avoids the "native-like" behaviour of RHF since [the current code](https://github.com/react-hook-form/react-hook-form/blob/68333911adee5ccc6fabd3966294a00ff296f569/src/logic/getFieldsValues.ts#L22-L24) only checks if `ref.disabled` is true. **Can this workaround be relied upon to work in the future as well?**

## Inconsistency between RHF and native `FormData` for disabled inputs
I also noticed what is likely a bug with the native-like behaviour of RHF. It doesn't seem to omit the values of disabled checkboxes and radio buttons, unlike native `FormData` behaviour:
<img width="295" alt="Screenshot 2021-03-12 at 12 21 13" src="https://user-images.githubusercontent.com/150084/110933894-9b778380-832d-11eb-8d97-a2cfd7e6e95b.png">

**Comment by bluebill1049** [comment]
reference we had a poll on this: https://twitter.com/HookForm/status/1359637785414488066

**Reply by mogelbrod** [reply]
I've seen it 👍 I also recall (perhaps incorrectly, I can't seem to find any reference to it now) that you some other time asked the community whether or not `disabled` inputs should be omitted - where the results indicated that most people preferred omitting the values IIRC (which is the current behaviour). I do think that poll is somewhat biased as there was no option to vote for `{ name: currentValue}` - but I think the reasoning behind the current behaviour is sound 😃 As the suggested workaround doesn't apply to the general case I do think it's worth noting though, and perhaps reconsider what the generally suggested (or "approved") approach is.

Alas that isn't my main question in this thread. Here's the (somewhat hidden) primary question I have:

> The workaround I found for this is to **wrap the elements with a `<fieldset>` and disable it during submit rather than the inputs themselves**. This can be tested in the linked CSB by selecting "disabled fieldset" in "Input state during submit". It avoids the "native-like" behaviour of RHF since [the current code](https://github.com/react-hook-form/react-hook-form/blob/68333911adee5ccc6fabd3966294a00ff296f569/src/logic/getFieldsValues.ts#L22-L24) only checks if `ref.disabled` is true. 
> ### Can this workaround be relied upon to work in the future as well?

If the answer is yes then I'm happy to simply wrap fields with a fieldset that I toggle `disabled` on - it has the theoretical advantage of avoiding input re-renders while toggling `disabled` anyway. If no then I'm struggling to find a way to prevent modifications to RHF forms during submissions 🙁

**Reply by bluebill1049** [reply]
personally, I have much preferred the native behavior which omits the value instead of `undefined` and definitely not retains the value. I guess the question is how much of impact would be to change over for omit behavior, also those users who have used with `undefined` and lead to one more breaking change from v6 to v7. I am happy to hear what everyone else's thinks as well. It's a rather difficult decision to be made by the library. eg will that impact how schema validation works from `undefined` to omit, but the good thing is we still have time before release the final v7 before it potentially becomes another regret.

**Reply by bluebill1049** [reply]
By the way, the easiest workaround would probably `unregister` that input, which probably easier to achieve on `useController`.

**Reply by mogelbrod** [reply]
Absolutely! As long as it's possible to work around the current behavior by wrapping inputs with `<fieldset disabled>` I'm 100% okay with it (which is currently the case) 👍It sounds like you have no plan of removing this "unintended" feature then? 🤞🙂 

> By the way, the easiest workaround would probably `unregister` that input, which probably easier to achieve on `useController`.

I'm unsure what you're suggesting here - is it to unregister all inputs while `formState.isSubmitting` is true and then register them again when it goes back to false? The form framework I'm developing supports both RHF as well as other form libraries, and adding this kind of coupling specifically for RHF would be a bit cumbersome.

**Reply by bluebill1049** [reply]
> It sounds like you have no plan of removing this "unintended" feature then?

I am investigating the impact while omitting the value (with disabled) eg: (I am lean toward omi value with disabled to align with HTLM standard)

- watch
- schema
- internal impact

and etc.

**Reply by bluebill1049** [reply]
I think we will go with omitting value and skip validation to align with HTML standard: 

https://codesandbox.io/s/react-hook-form-readonly-vs-disabled-forked-fsbiq?file=/src/index.js

**Reply by finnhvman** [reply]
I want to disable my inputs when the form is submitting, and if the form submits successfully (e.g. it's valid) there are no issues. However, if I have cross field validation where one field is validated based on the value of another field, it can cause issues. Say the form isn't valid, user clicks submit, then the cross field validation breaks because at this point the watched value of the "depended on" field becomes `undefined`.

I think it's a common pattern to disable fields while submitting, but I don't think a breaking change is necessary to solve this issue. How about introducing a new boolean option to `useForm`, where we can opt out of disabled fields being omitted?

**Reply by mogelbrod** [reply]
> I think we will go with omitting value and skip validation to align with HTML standard:
> 
> https://codesandbox.io/s/react-hook-form-readonly-vs-disabled-forked-fsbiq?file=/src/index.js

In this fork the `<fieldset disabled>` workaround still works, which is great. The fieldset workaround should then continue being usable for users who want disabled inputs to retain their values, yes?

BTW In this fork the checkbox & radio buttons are hardcoded to be disabled at all times, was that intentional?

> I think it's a common pattern to disable fields while submitting, but I don't think a breaking change is necessary to solve this issue. How about introducing a new boolean option to `useForm`, where we can opt out of disabled fields being omitted?

While any addition of new configuration options increase the risk of confusing the library users, I do think it may be reasonable to consider in this case as the possible two behaviours (include vs exclude `<input disabled>` values) both have their own advantages and disadvantages. An opt-in flag like `includeDisabled`/`includeDisabledValues` could improve developer experience as it also clearly communicates that the default behaviour is to omit the values.

Personally, as long as values for inputs within `<fieldset disabled>` are still included I don't see this as a must have - but if that option was removed then I'd definitely request this kind of configuration flag.

**Reply by bluebill1049** [reply]
That was one of the reasons why disabled input was return undefined instead of omitting the value because developers want to have different behavior as the web standard. I wouldn't introduce another config for `disabled`, we either set it as undefined or omit the value. If users want to have custom behavior do at the controlled component level.

**Reply by bluebill1049** [reply]
> Say the form isn't valid, user clicks submit, then the cross field validation breaks because at this point the watched value of the "depended on" field becomes `undefined`.

It should return undefined and it's already doing.

**Reply by bluebill1049** [reply]
`disabled` should be: If a field is disabled, the value of the field is not sent to the server when the form is submitted. 

if we align with the standard then the value should be omitted, if we getting closer to what React developers' expected will be `undefiend`. I will most likely make a decision by today and we either keep what we have `undefiend` or moving on to omitted and off course documented this down.

**Reply by finnhvman** [reply]
What is the difference between the value being omitted and the value being `undefined`? Aren't they the same thing?

**Reply by bluebill1049** [reply]
> What is the difference between the value being omitted and the value being `undefined`? Aren't they the same thing?

{
  test: undefined
}

vs 

{}

**Reply by bluebill1049** [reply]
The decision is made to keep what we have right now which returns `{ test: undefined }` with disabled input "test". I am also patching the two issues with the checkbox and radio input.

**Reply by bluebill1049** [reply]
here is the relevant PR: https://github.com/react-hook-form/react-hook-form/pull/4422

**Reply by mogelbrod** [reply]
Nice to see the bug fixed!

I just want to clarify this:
> An opt-in flag like `includeDisabled`/`includeDisabledValues` could improve developer experience as it also clearly communicates that the default behaviour is to omit the values.

I'm not referring to `{ test: undefined }` vs `{}` but rather an option that switches RHF to include the current value of all disabled fields:
```jsx
const form = useForm({
  includeDisabledValues: true,
})
const renderedField = <input disabled {...form.register('field') } />
// Enter value into input at some point (normally done by user input)
someRefToRenderedField.current.value = 'abc'
// Get values:
form.getValues() // should return: { field: 'abc' }
```

But as I've mentioned before such configuration flag would only be necessary if you at some point make inputs disabled via `<fieldset disabled>` not have their current values included in `getValues()`. That is, if RHF was changed so that the CSB instead gave this result **(screenshot 1)**:
<img width="297" alt="Screenshot 2021-03-15 at 09 29 05" src="https://user-images.githubusercontent.com/150084/111133610-fe118f00-857a-11eb-99ce-5c9e0a8347fc.png">

The current behaviour on the other hand retains values for inputs within `<fieldset disabled>` **(screenshot 2)**:
<img width="297" alt="Screenshot 2021-03-15 at 09 29 05" src="https://user-images.githubusercontent.com/150084/111124653-095fbd00-8571-11eb-915a-ae018f53c8ec.png">

The current behaviour is great as it makes it somewhat simple to disable the form inputs without the values being affected. However, this doesn't seem align with your intention of making RHF follow the DOM behaviour since `FormData` omits inputs disabled via an ancestor `<fieldset disabled>`.

**So just to clear up any confusion:** Do you plan on keeping the current behaviour showcased in **screenshot 2** (which is what I'm hoping to be the case)? Or do you intend to modify RHF to make it adhere to the `FormData` behaviour showcased in **screenshot 1**?

If screenshot 2 will continue being the case then my questions in this thread can be considered answered/solved 😃

**Reply by bluebill1049** [reply]
thanks for the clarification. keep it simple and consistent: 

- disabled input will return `undefined`.

here is an updated CSB: https://codesandbox.io/s/react-hook-form-readonly-vs-disabled-forked-sxgye?file=/src/index.js

**Reply by mogelbrod** [reply]
May I suggest the following rephrasing?

> Form inputs with the `disabled` attribute will have value `undefined`.

Since inputs disabled by an ancestor `<fieldset disabled>` are semantically "disabled inputs" this phrasing should help disambiguate between the two.

**Reply by bluebill1049** [reply]
Will include that in the doc.

**Reply by harveyappleton** [reply]
Hey @bluebill1049 ! thanks once again for an awesome library, we love RHF.

Quick question. I noticed this became an issue for us when we updated to v7. 

We use Antd components, and they do not have a `readOnly` prop, only `disabled`, and the `fieldset` solution doesn't work for them either.

We want to be able to disable inputs/switches/etc, but retain their value. 
For example:

i/o = MAIN ENABLED SWITCH

i/o = switch 1
i/o = switch 2
...etc

We'd like to be able to submit something like 
```
{
mainSwitch: false,
switch1: false,
switch2: true
...
}
```

(of which switch 1, 2.. etc are disabled when main switch is changed).

Any advice?

Thanks!

**Reply by bluebill1049** [reply]
hey @harveyappleton what about this:

```
<fieldset disabled><input /></fieldset>
```

**Reply by harveyappleton** [reply]
Hey @bluebill1049 - we use Antd Inputs (see [https://ant.design/components/input/](https://ant.design/components/input/)) - they are React components that do not work with a wrapped `<fieldset>` I'm afraid :( Any other good workarounds? This is causing a problem if I upgrade to latest version sadly.

**Reply by functor-dev** [reply]
Did you figure out @harveyappleton?

**Comment by dextermb** [comment]
I was hoping to find that a `readOnly` input's value is sent to the server, but disabled is not, unfortunately it seems both readOnly` and `disabled` do not send values to the server.

How I see `readOnly` is that it should be used for a pre-populated value so the client can see the value but not edit it.
